src/python/primtall/primtall.md

```diff
@@ -0,0 +1,195 @@
+---
+title: Primtall og effektivitet
+level: 3
+author: Martin Strand
+language: nb
+tags:
+  topic: [text_based]
+  subject: [mathematics, programming, cryptography]
+  grade: []
+---
+
+# Introduksjon {.intro}
+
+I matematikktimene har du lært om primtall, tall som er slik at de bare kan deles av seg selv og 1. Nå skal vi se hvordan vi kan undersøke om et vilkårlig heltall er et primtall eller ikke.
+
+Når vi skal kryptere data er vi ofte avhengige av å jobbe med store primtall, og vi vil få bruk for det i senere oppgaver. For å jobbe med primtall, må vi først ha en måte å finne dem på.
+
+# Steg 1: Rest etter divisjon {.activity}
+
+Den grunnleggende måten vi gjør dette på er å dele på andre tall. La oss si at vi vil sjekke tallet 15. Tallet 2 går ikke opp i 15, så da sier vi at vi får en *rest*. Åpne IDLE, men uten å lage en ny fil. Der det står `>>>` kan vi skrive inn kode for å se hva den gjør.
+
+## Sjekkliste {.check}
+
+- [ ] Skriv `15//2` og trykk Enter. Du skal da få svaret 7. Det er fordi 2 går opp 7 ganger i 15
+- [ ] Skriv `15 % 2` og trykk Enter. Da skal du få 1. Til sammen betyr dette at 15 delt på 2 blir 7, med 1 til rest.
+- [ ] Prøv det samme med `15//6` og `15//6`. Hva betyr svarene du får ut?
+- [ ] Skriv `15 % 3`. Da får du 0. Hvorfor?
+
+# Steg 2: Enkel prøvedivisjon {.activity}
+
+Over fikk vi 0 når vi regnet ut `15 % 3`. Det betyr at 3 går opp i 15, og altså er ikke 15 et primtall. Det kan vi utnytte for å sjekke om noe er et primtall. Vi kan dermed lage en funksjon som tar inn et tall `n` og sjekker om alle tall som er mindre enn `n` går opp i `n`.
+
+## Sjekkliste {.check}
+
+- [ ] Lag en ny fil med følgende kode
+
+  ```python
+  def is_prime(n):
```

Comment:
sjekke at input er positivt heltall og ikke like 0 eller 1 osv? kan jo også være verdt en kommentar angående sikker programmering at man alltid må verifisere at input er slik det burde være. (dette er jo også grunn til at flere kryptografiske protokoller har blitt brutt opp gjennom årene)

Reply:
Godt poeng. Jeg greier ikke helt å bestemme meg her: Vi vil jo gjerne lære bort god programmeringsskikk, så en test av input burde absolutt være med. Samtidig kan jo et slikt tillegg med dets forklaring ta fokus bort fra de andre tingene vi vil ha fram, og dermed gjøre terskelen litt større. Hva tror du?

(Om jeg legger det inn, blir det if not (isinstance(n, (int, long)) and n > 1): <break> return False. Det er litt teknisk, men mindre teknisk enn å sende en exception, som er det man burde gjøre ...)

Reply:
synes if not (isinstance(n, (int, long)) and n > 1): return False er en god løsning her, som ikke blir for avansert eller tar for mye fokus bort fra hovedtemaet i oppgaven. legg gjerne til dette + en komentar om hva du gjør og hvorfor.